AppSyncRealTimeClient/ConnectionProvider/AppsyncRealtimeConnection/RealtimeConnectionProvider.swift

```diff
@@ -93,15 +94,26 @@ public class RealtimeConnectionProvider: ConnectionProvider {
     }
 
     public func addListener(identifier: String, callback: @escaping ConnectionProviderCallback) {
-        serialCallbackQueue.async {
-            self.listeners[identifier] = callback
+        serialCallbackQueue.async { [weak self] in
+            self?.listeners[identifier] = callback
         }
-
     }
 
     public func removeListener(identifier: String) {
-        serialCallbackQueue.async {
+        serialCallbackQueue.async { [weak self] in
+            guard let self = self else {
+                return
+            }
+
             self.listeners.removeValue(forKey: identifier)
+
+            self.serialConnectionQueue.async { [weak self] in
+                guard let self = self else {
+                    return
+                }
+                self.status = .notConnected
```

Comment:
Shouldnt we check the number of connections here before sending disconnect to the websocket?

Reply:
woops, missed this after moving implemenentions back and forth